prov/efa/src/efa_cq.h

```diff
@@ -176,3 +184,111 @@ static inline int efa_cq_ibv_cq_ex_open(struct fi_cq_attr *attr,
+ * @return true the wc consumes a recv buffer
+ * @return false the wc doesn't consume a recv buffer
+ */
+static inline
+bool efa_cq_wc_is_unsolicited(struct ibv_cq_ex *ibv_cq_ex)
+{
+	return efa_use_unsolicited_write_recv() && efadv_wc_is_unsolicited(efadv_cq_from_ibv_cq_ex(ibv_cq_ex));
+}
+
+#else
+
+static inline
+bool efa_cq_wc_is_unsolicited(struct ibv_cq_ex *ibv_cq_ex)
+{
+	return false;
+}
+
+#endif
+
+static inline const char *efa_ep_raw_addr_str(struct efa_base_ep *base_ep, char *buf, size_t *buflen)
+{
+	return ofi_straddr(buf, buflen, FI_ADDR_EFA, &base_ep->src_addr);
+}
+
+static inline const char *efa_ep_get_peer_raw_addr_str(struct efa_base_ep *base_ep, fi_addr_t addr, char *buf, size_t *buflen)
+{
+	struct efa_conn *efa_conn;
+	
+	if (OFI_UNLIKELY(addr == FI_ADDR_NOTAVAIL))
+		efa_conn = NULL;
+	else
+		efa_conn = efa_av_addr_to_conn(base_ep->av, addr);
+
+	return ofi_straddr(buf, buflen, FI_ADDR_EFA, efa_conn ? efa_conn->ep_addr : NULL);
+}
+
+/**
+ * @brief Write the error message and return its byte length
+ * @param[in]    ep          EFA base endpoint
+ * @param[in]    addr        Remote peer fi_addr_t
+ * @param[in]    prov_errno  EFA provider * error code(must be positive)
+ * @param[out]   buf         Pointer to the address of error data written by
+ * this function
+ * @param[out]   buflen      Pointer to the returned error data size
+ * @return       A status code. 0 if the error data was written successfully,
+ * otherwise a negative FI error code.
+ */
+static inline int efa_write_error_msg(struct efa_base_ep *ep, fi_addr_t addr,
+				      int prov_errno, void **buf,
+				      size_t *buflen)
+{
+	char ep_addr_str[OFI_ADDRSTRLEN] = {0}, peer_addr_str[OFI_ADDRSTRLEN] = {0};
+	char peer_host_id_str[EFA_HOST_ID_STRING_LENGTH + 1] = {0};
+	char local_host_id_str[EFA_HOST_ID_STRING_LENGTH + 1] = {0};
+	const char *base_msg = efa_strerror(prov_errno);
+	char err_msg[EFA_ERROR_MSG_BUFFER_LENGTH];
+	size_t len = 0;
+	uint64_t local_host_id;
+
+	*buf = NULL;
+	*buflen = 0;
+
+	len = sizeof(ep_addr_str);
+	efa_ep_raw_addr_str(ep, ep_addr_str, &len);
+	len = sizeof(peer_addr_str);
+	efa_ep_get_peer_raw_addr_str(ep, addr, peer_addr_str, &len);
+
+	local_host_id = efa_get_host_id(efa_env.host_id_file);
+	if (!local_host_id ||
+	    EFA_HOST_ID_STRING_LENGTH != snprintf(local_host_id_str,
+						  EFA_HOST_ID_STRING_LENGTH + 1,
+						  "i-%017lx", local_host_id)) {
+		strcpy(local_host_id_str, "N/A");
+	}
+
+	/* efa-raw cannot get peer host id without a handshake */
+	strcpy(peer_host_id_str, "N/A");
+
+	int ret = snprintf(err_msg, EFA_ERROR_MSG_BUFFER_LENGTH,
+			   "%s My EFA addr: %s My host id: %s Peer EFA addr: "
+			   "%s Peer host id: %s",
+			   base_msg, ep_addr_str, local_host_id_str,
+			   peer_addr_str, peer_host_id_str);
+
+	if (ret < 0 || ret > EFA_ERROR_MSG_BUFFER_LENGTH - 1) {
+		return -FI_EINVAL;
+	}
+
+	if (strlen(err_msg) >= EFA_ERROR_MSG_BUFFER_LENGTH) {
+		return -FI_ENOBUFS;
+	}
+
+	*buf = err_msg;
```

Comment:
Are you sure this will work? Your `err_msg` will be released when the function returns as it is declared within the stack.

Reply:
I guess it might cause undefined behavior. I will move err_msg to be a field of base_ep then.

Reply:
Do you have to use a persistent `err_msg` in `base_ep` ? If you read the `ofi_cq_write_err` code, you will find it will `mem_dup` the `err_data` you filled in `entry`. In other words, the `err_data` you passed can be safely released after `ofi_cq_write_err` returns. Just use a temporary char array in your code block that write the error cq.